cpp/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -1105,21 +1162,24 @@ using BlockMap = std::unordered_map<int, std::shared_ptr<PandasBlock>>;
 // * placement arrays as we go
 class DataFrameBlockCreator {
  public:
-  explicit DataFrameBlockCreator(const std::shared_ptr<Table>& table) : table_(table) {}
+  explicit DataFrameBlockCreator(PandasOptions options,
+                                 const std::shared_ptr<Table>& table, MemoryPool* pool)
+      : table_(table), options_(options), pool_(pool) {}
 
   Status Convert(int nthreads, PyObject** output) {
     column_types_.resize(table_->num_columns());
     column_block_placement_.resize(table_->num_columns());
     type_counts_.clear();
     blocks_.clear();
 
-    RETURN_NOT_OK(CreateBlocks());
+    RETURN_NOT_OK(InferOutputBlockTypes());
+    RETURN_NOT_OK(CreateTrivialBlocks());
```

Comment:
- I don't think that you can overwrite `PandasBlock::Allocate()` in a meaningful way without templating.
- Since the Encode->Allocate->Write order can not be changed I think putting all of it into a single function makes sense. What's slightly bothering me is that the CategoricalBlock would behave very differently compared to the other Blocks if we wrap all of this into the Write function. Yet, it is already somehow special and we might need to take that bullet...